File: go800mon/a800mon/screenbuffer.go
package a800mon

import (
	"context"
	"encoding/binary"
	"fmt"
	"hash/fnv"
	"strings"

	"go800mon/internal/atascii"
	dl "go800mon/internal/displaylist"
)

type ScreenBufferInspector struct {
	BaseWindowComponent
	rpc            *RpcClient
	grid           *GridWidget
	screen         *Screen
	lastUseATASCII bool
	lastSnapshot   string
}

type rowRangeIndex struct {
	s   int
	e   int
	off int
}

func NewScreenBufferInspector(rpc *RpcClient, window *Window) *ScreenBufferInspector {
	grid := NewGridWidget(window)
	grid.SetGridColumnGap(0)
	grid.SetGridSelectionEnabled(false)
	return &ScreenBufferInspector{
		BaseWindowComponent: NewBaseWindowComponent(window),
		rpc:                 rpc,
		grid:                grid,
	}
}

func (s *ScreenBufferInspector) BindInput(screen *Screen) {
	s.screen = screen
}

func (s *ScreenBufferInspector) HandleInput(ch int) bool {
	if State().InputFocus {
		return false
	}
	if s.screen == nil || !(s.screen.Focused() == s.Window()) {
		return false
	}
	if !(ch == int(' ') || ch == int('a') || ch == int('A')) {
		return s.grid.HandleGridNavigationInput(ch)
	}
	store.setUseATASCII(!State().UseATASCII)
	return true
}

func (s *ScreenBufferInspector) Update(ctx context.Context) (bool, error) {
	if State().InputFocus {
		return false, nil
	}
	st := State()
	mapper := dl.NewMemoryMapper(st.DList, st.DMACTL, 0x400)
	fetchRanges, rowSlices := mapper.Plan()
	segs := st.DList.ScreenSegments(st.DMACTL)
	if st.DListSelectedRegion != nil && *st.DListSelectedRegion >= 0 && *st.DListSelectedRegion < len(segs) {
		seg := segs[*st.DListSelectedRegion]
		fetchRanges = []dl.FetchRange{{Start: seg.Start, End: seg.End}}
		filtered := make([]dl.RowSlice, 0, len(rowSlices))
		for _, r := range rowSlices {
			addr := int(r.Addr)
			if seg.Start <= addr && addr < seg.End {
				filtered = append(filtered, r)
			}
		}
		rowSlices = filtered
	}
	if len(fetchRanges) == 0 {
		store.setScreenRows(nil)
		if s.lastSnapshot == "" {
			return false, nil
		}
		s.lastSnapshot = ""
		return true, nil
	}
	buffer := make([]byte, 0, 8192)
	index := make([]rowRangeIndex, 0, len(fetchRanges))
	off := 0
	for _, r := range fetchRanges {
		ln := r.End - r.Start
		if ln <= 0 {
			continue
		}
		chunk, err := s.rpc.ReadMemoryChunked(ctx, uint16(r.Start&0xFFFF), ln)
		if err != nil {
			return false, nil
		}
		buffer = append(buffer, chunk...)
		index = append(index, rowRangeIndex{s: r.Start, e: r.End, off: off})
		off += len(chunk)
	}
	rows := make([]ScreenRow, 0, len(rowSlices))
	for _, rs := range rowSlices {
		row := readRow(buffer, index, int(rs.Addr), rs.Length)
		if len(row) == 0 {
			continue
		}
		rows = append(rows, ScreenRow{Addr: rs.Addr, Data: row})
	}
	store.setScreenRows(rows)
	snapshot := buildScreenRowsSnapshot(rows)
	if s.lastSnapshot == snapshot {
		return false, nil
	}
	s.lastSnapshot = snapshot
	return true, nil
}

func readRow(buffer []byte, index []rowRangeIndex, addr, ln int) []byte {
	if ln <= 0 {
		return nil
	}
	out := make([]byte, 0, ln)
	cur := addr
	remaining := ln
	for remaining > 0 {
		found := false
		for _, it := range index {
			if it.s <= cur && cur < it.e {
				take := it.e - cur
				if take > remaining {
					take = remaining
				}
				from := it.off + (cur - it.s)
				to := from + take
				if from < 0 || to > len(buffer) || from >= to {
					return nil
				}
				out = append(out, buffer[from:to]...)
				cur += take
				remaining -= take
				found = true
				break
			}
		}
		if !found {
			return nil
		}
	}
	return out
}

func (s *ScreenBufferInspector) Render(_force bool) {
	st := State()
	w := s.Window()
	if s.lastUseATASCII != st.UseATASCII {
		s.lastUseATASCII = st.UseATASCII
		w.SetTagActive("atascii", st.UseATASCII)
		w.SetTagActive("ascii", !st.UseATASCII)
	}
	contentWidth := w.Width() - 8
	if contentWidth < 0 {
		contentWidth = 0
	}
	rows := make([]ScreenRow, 0, w.Height())
	drawWidth := 0
	for _, row := range st.ScreenRows {
		if len(row.Data) > contentWidth {
			row.Data = row.Data[:contentWidth]
		}
		if len(row.Data) > drawWidth {
			drawWidth = len(row.Data)
		}
		rows = append(rows, row)
	}
	if drawWidth > contentWidth {
		drawWidth = contentWidth
	}
	gridRows := make([]GridRow, 0, len(rows))
	for _, row := range rows {
		rowLen := len(row.Data)
		if rowLen > drawWidth {
			rowLen = drawWidth
		}
		leftPad := 0
		rightPad := 0
		if drawWidth > rowLen {
			leftPad = (drawWidth - rowLen) / 2
			rightPad = drawWidth - rowLen - leftPad
		}
		cells := GridRow{{Text: formatHex16(row.Addr) + ": ", Attr: ColorAddress.Attr()}}
		if leftPad > 0 {
			cells = append(cells, GridCell{Text: strings.Repeat("·", leftPad), Attr: ColorUnused.Attr()})
		}
		cells = append(cells, renderScreenRuns(row.Data[:rowLen], st.UseATASCII)...)
		if rightPad > 0 {
			cells = append(cells, GridCell{Text: strings.Repeat("·", rightPad), Attr: ColorUnused.Attr()})
		}
		gridRows = append(gridRows, cells)
	}
	s.grid.SetGridColumnWidths(nil)
	s.grid.SetGridRows(gridRows)
	s.grid.SetGridSelected(nil)
	s.grid.RenderGrid()
}

func renderScreenRuns(data []byte, useATASCII bool) []GridCell {
	if len(data) == 0 {
		return nil
	}
	if !useATASCII {
		out := make([]rune, 0, len(data))
		for _, b := range data {
			if b <= 0 {
				out = append(out, ' ')
				continue
			}
			v := b & 0x7F
			if v >= 32 && v <= 126 {
				out = append(out, rune(v))
			} else {
				out = append(out, '.')
			}
		}
		return []GridCell{{Text: string(out), Attr: ColorText.Attr()}}
	}
	runs := make([]GridCell, 0, len(data))
	buf := make([]rune, 0, len(data))
	curAttr := -1
	flush := func() {
		if len(buf) == 0 || curAttr < 0 {
			return
		}
		runs = append(runs, GridCell{Text: string(buf), Attr: ColorText.Attr() | curAttr})
		buf = buf[:0]
	}
	for _, b := range data {
		ch, attr := renderScreenCharATASCII(b)
		if curAttr < 0 {
			curAttr = attr
		}
		if attr != curAttr {
			flush()
			curAttr = attr
		}
		buf = append(buf, ch)
	}
	flush()
	return runs
}

func renderScreenCharATASCII(b byte) (rune, int) {
	if b <= 0 {
		return ' ', 0
	}
	a := atascii.ScreenToATASCII(b)
	r := []rune(atascii.LookupPrintable(a & 0x7F))
	ch := '.'
	if len(r) > 0 {
		ch = r[0]
	}
	attr := 0
	if (a & 0x80) != 0 {
		attr = AttrReverse()
	}
	return ch, attr
}

func buildScreenRowsSnapshot(rows []ScreenRow) string {
	h := fnv.New64a()
	var n [2]byte
	binary.LittleEndian.PutUint16(n[:], uint16(len(rows)))
	_, _ = h.Write(n[:])
	for _, r := range rows {
		binary.LittleEndian.PutUint16(n[:], r.Addr)
		_, _ = h.Write(n[:])
		binary.LittleEndian.PutUint16(n[:], uint16(len(r.Data)))
		_, _ = h.Write(n[:])
		if len(r.Data) > 0 {
			_, _ = h.Write(r.Data)
		}
	}
	return fmt.Sprintf("%016X", h.Sum64())
}

func formatHex16(v uint16) string {
	const hex = "0123456789ABCDEF"
	b := [4]byte{hex[(v>>12)&0xF], hex[(v>>8)&0xF], hex[(v>>4)&0xF], hex[v&0xF]}
	return string(b[:])
}

File: py800mon/screenbuffer.py
import dataclasses

from . import debug
from .app import VisualRpcComponent
from .appstate import state, store
from .atascii import atascii_to_curses, screen_to_atascii
from .datastructures import ScreenBuffer
from .displaylist import DMACTL_ADDR, DMACTL_HW_ADDR, DisplayListMemoryMapper
from .rpc import RpcException
from .ui import Color, GridCell, GridWidget


@dataclasses.dataclass(frozen=True, slots=True)
class ScreenBufferCell:
    addr: int  # adres w Twoim buforze (np. 0x3800 + offset)
    value: int  # bajt 0..255 (wartość "ekranowa" / ATASCII / to co trzymasz)
    x: int
    y: int

    def as_int(self) -> int:
        return self.value

    def as_ascii(self) -> str:
        # Minimalnie: klasyczne 7-bit ASCII (kontrola -> '.')
        v = self.value & 0x7F
        if 32 <= v <= 126:
            return chr(v)
        return "."

    def as_atascii(self) -> int:
        # Zwraca "surową" wartość ATASCII (jak trzymasz w buforze)
        return self.value & 0xFF

    def as_atascii_char(self) -> str:
        # Jeśli masz własne mapowanie ATASCII->Unicode, podepnij je tutaj.
        # Domyślnie: ASCII z maską 0x7F.
        return self.as_ascii()


class ScreenBufferInspector(VisualRpcComponent):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.grid = GridWidget(self.window, col_gap=0)
        self.grid.set_grid_selection_enabled(False)
        self._inspect = False
        self._cx = 0
        self._cy = 0
        self._dmactl = 0
        self._last_use_atascii = None
        self._screen = None

    def bind_input(self, screen):
        self._screen = screen

    @property
    def cols(self):
        return self.window.w

    @property
    def rows(self):
        return self.window.h

    def toggle_inspect(self) -> bool:
        self._inspect = not self._inspect
        return self._inspect

    def set_cursor(self, x: int, y: int) -> None:
        x = int(x)
        y = int(y)

        if self.cols <= 0 or self.rows <= 0:
            return

        # clamp
        if x < 0:
            x = 0
        elif x >= self.cols:
            x = self.cols - 1

        if y < 0:
            y = 0
        elif y >= self.rows:
            y = self.rows - 1

        self._cx = x
        self._cy = y

    def cursor_left(self, n: int = 1) -> None:
        self.set_cursor(self._cx - int(n), self._cy)

    def cursor_right(self, n: int = 1) -> None:
        self.set_cursor(self._cx + int(n), self._cy)

    def cursor_up(self, n: int = 1) -> None:
        self.set_cursor(self._cx, self._cy - int(n))

    def cursor_down(self, n: int = 1) -> None:
        self.set_cursor(self._cx, self._cy + int(n))

    def is_inspecting(self) -> bool:
        return self._inspect

    def handle_input(self, ch):
        if state.input_focus:
            return False
        if self._screen is None or not self._screen.focused == self.window:
            return False
        if ch in (ord(" "), ord("a"), ord("A")):
            store.set_use_atascii(not state.use_atascii)
            return True
        return self.grid.handle_grid_navigation_input(ch)

    @property
    def cursor(self) -> tuple[int, int]:
        return self._cx, self._cy

    @property
    def cell(self) -> ScreenBufferCell:
        val = self.window.get_char(self._cx, self._cy) & 0xFF
        return ScreenBufferCell(addr=0, value=val, x=self._cx, y=self._cy)

    def put_value(self, value: int) -> None:
        raise NotImplementedError

    def render(self, force_redraw=False) -> None:
        if self._last_use_atascii != state.use_atascii:
            self._last_use_atascii = state.use_atascii
            self.window.set_tag_active("atascii", state.use_atascii)
            self.window.set_tag_active("ascii", not state.use_atascii)

        segs = state.dlist.screen_segments(self._dmactl)
        active_seg = None
        if state.dlist_selected_region is not None and segs:
            if 0 <= state.dlist_selected_region < len(segs):
                active_seg = segs[state.dlist_selected_region]
        content_width = self.window._iw - 8
        if content_width < 0:
            content_width = 0
        rows_to_draw = []
        for row_info in state.screen_buffer.row_slices:
            parsed = _parse_row_info(row_info, state.screen_buffer.start_address)
            if not parsed:
                continue
            start_addr, length = parsed
            if active_seg is not None:
                start, end, _mode = active_seg
                if not (start <= start_addr < end):
                    continue
            if length <= 0:
                continue
            row = state.screen_buffer.get_range(start_addr, length)[:content_width]
            rows_to_draw.append((start_addr, row))
        draw_width = 0
        for _addr, row in rows_to_draw:
            if len(row) > draw_width:
                draw_width = len(row)
        if draw_width > content_width:
            draw_width = content_width

        rows = []
        for start_addr, row in rows_to_draw:
            row = row[:draw_width]
            left_pad = 0
            right_pad = 0
            if draw_width > len(row):
                left_pad = (draw_width - len(row)) // 2
                right_pad = draw_width - len(row) - left_pad
            cells = [GridCell(f"{start_addr:04X}: ", Color.ADDRESS.attr())]
            if left_pad > 0:
                cells.append(GridCell("·" * left_pad, Color.UNUSED.attr()))
            for text, attr in _render_runs(row, state.use_atascii):
                if text:
                    cells.append(GridCell(text, attr))
            if right_pad > 0:
                cells.append(GridCell("·" * right_pad, Color.UNUSED.attr()))
            rows.append(tuple(cells))

        self.grid.set_grid_column_widths(())
        self.grid.set_grid_rows(rows)
        self.grid.set_grid_selected(None)
        self.grid.render_grid()

    async def update(self):
        try:
            dmactl = await self.rpc.read_byte(DMACTL_ADDR)
            if (dmactl & 0x03) == 0:
                dmactl = await self.rpc.read_byte(DMACTL_HW_ADDR)
            mapper = DisplayListMemoryMapper(state.dlist, dmactl)
            fetch_ranges, row_slices = mapper.plan()
            segs = state.dlist.screen_segments(dmactl)
            if (
                state.dlist_selected_region is not None
                and 0 <= state.dlist_selected_region < len(segs)
            ):
                seg_start, seg_end, _mode = segs[state.dlist_selected_region]
                fetch_ranges = [(seg_start, seg_end)]
                row_slices = [
                    (addr, length)
                    for addr, length in mapper.row_ranges()
                    if addr is not None and seg_start <= addr < seg_end
                ]
            chunks = []
            for s, e in fetch_ranges:
                chunks.append(await self.rpc.read_memory(s, e - s))
            buffer = b"".join(chunks)
            start_address = fetch_ranges[0][0] if fetch_ranges else 0
            range_index = []
            offset = 0
            for s, e in fetch_ranges:
                range_index.append((s, e, offset))
                offset += e - s
            store.set_screen_buffer(
                ScreenBuffer(
                    row_slices=row_slices,
                    buffer=buffer,
                    start_address=start_address,
                    range_index=range_index,
                )
            )
            self._dmactl = dmactl
            return True
        except RpcException:
            return False


def _render_char(value: int, use_atascii: bool):
    if value <= 0:
        return " ", 0
    if use_atascii:
        return atascii_to_curses(screen_to_atascii(value))
    v = value & 0x7F
    if 32 <= v <= 126:
        return chr(v), 0
    return ".", 0


def _render_runs(row: bytes, use_atascii: bool):
    if not row:
        return []
    if not use_atascii:
        text = []
        for b in row:
            if b <= 0:
                text.append(" ")
                continue
            v = b & 0x7F
            if 32 <= v <= 126:
                text.append(chr(v))
            else:
                text.append(".")
        return [("".join(text), 0)]

    runs = []
    cur_attr = None
    buf = []
    for b in row:
        ch, attr = _render_char(b, use_atascii=True)
        if cur_attr is None:
            cur_attr = attr
        if attr != cur_attr:
            runs.append(("".join(buf), cur_attr))
            buf = [ch]
            cur_attr = attr
            continue
        buf.append(ch)
    if buf:
        runs.append(("".join(buf), cur_attr))
    return runs


def _parse_row_info(row_info, base_addr: int):
    if row_info is None:
        return None
    if isinstance(row_info, tuple):
        if len(row_info) < 2:
            return None
        first, second = row_info[0], row_info[1]
        if isinstance(first, slice):
            if first.start is None or first.stop is None or second is None:
                return None
            return int(second), int(first.stop - first.start)
        if first is None or second is None:
            return None
        return int(first), int(second)
    if isinstance(row_info, slice):
        if row_info.start is None or row_info.stop is None:
            return None
        return int(base_addr + row_info.start), int(row_info.stop - row_info.start)
    return None
